WalletWasabi.Gui/TransactionWatcher.cs

```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.InteropServices;
+using WalletWasabi.Models;
+
+namespace WalletWasabi.Gui
+{
+	class CoinEventArgs : EventArgs
+	{
+		public SmartCoin Coin { get; }
+
+		public CoinEventArgs(SmartCoin coin)
+		{
+			Coin = coin;
+		}
+	}
+
+	class CoinsWatcher
+	{
+		private List<SmartCoin> _trackedCoins;
+		private bool _isSynchronized = false;
+		public EventHandler<CoinEventArgs> CoinReceived;
+
+		public CoinsWatcher()
+		{
+			_trackedCoins = new List<SmartCoin>();
+			OnBestHeightChanged(this, Height.Unknown);
+		}
+
+		public void Start()
+		{
+			Global.IndexDownloader.BestHeightChanged += OnBestHeightChanged;
+			Global.WalletService.Coins.HashSetChanged += OnCoinsArrives;
+		}
+
+		public void Stop()
+		{
+			Global.IndexDownloader.BestHeightChanged -= OnBestHeightChanged;
+			Global.WalletService.Coins.HashSetChanged -= OnCoinsArrives;
+		}
+		private void OnBestHeightChanged(object sender, Height e)
+		{
+			var filterLeft = Global.IndexDownloader.GetFiltersLeft();
+			if(filterLeft == 0)
+			{
+				Global.IndexDownloader.BestHeightChanged -= OnBestHeightChanged;
+				_trackedCoins = Global.WalletService.Coins.ToList();
+				_isSynchronized = true;
+			}
+		}
+
+		private void OnCoinsArrives(object sender, EventArgs e)
+		{
+			var allCoins = Global.WalletService.Coins;
+			
+			foreach(var coin in allCoins)
+			{
+				if( !_trackedCoins.Contains(coin) && _isSynchronized)
+				{
+					CoinReceived?.Invoke(this, new CoinEventArgs(coin));
+				}
+			}
+			_trackedCoins = allCoins.ToList();
+		}
+	}
+
+	class Notifier
+	{
+		private CoinsWatcher _watcher;
+
+		public static readonly Notifier Current = new Notifier();
+
+		private Notifier()
+		{
+			_watcher = new CoinsWatcher();
+		}
+
+		public void Start()
+		{
+			_watcher.Start();
+			_watcher.CoinReceived += NotifyIncomingCoin;
+		}
+
+		public void Stop()
+		{
+			_watcher.CoinReceived -= NotifyIncomingCoin;	
+			_watcher.Stop();
+		}
+
+		private void NotifyIncomingCoin(object sender, CoinEventArgs e)
```

Comment:
Exceptions here are going to be swallowed. Can you encapsulate with a `try catch { Logger.LogWarning }`?